lib/msprime.c

```diff
@@ -726,7 +882,71 @@ static segment_t *MSP_WARN_UNUSED
 msp_copy_segment(msp_t *self, const segment_t *seg)
 {
     return msp_alloc_segment(self, seg->left, seg->right, seg->value, seg->population,
-        seg->label, seg->prev, seg->next);
+        seg->label, seg->prev, seg->next, seg->hull);
+}
+
+static hull_t *MSP_WARN_UNUSED
+msp_alloc_hull(msp_t *self, double left, double right, segment_t *lineage)
+{
+    hull_t *hull = NULL;
+    label_id_t label;
+    uint32_t j;
+
+    label = lineage->label;
+
+    if (object_heap_empty(&self->hull_heap[label])) {
+        if (object_heap_expand(&self->hull_heap[label]) != 0) {
+            goto out;
+        }
+        /* check pointer logic here */
+        for (j = 0; j < self->num_populations; j++) {
+            if (self->populations[j].coal_mass_index != NULL) {
+                if (fenwick_expand(&self->populations[j].coal_mass_index[label],
+                        self->hull_block_size)
+                    != 0) {
+                    goto out;
+                }
+            }
+        }
+    }
+    hull = (hull_t *) object_heap_alloc_object(&self->hull_heap[label]);
+    if (hull == NULL) {
+        goto out;
+    }
+    tsk_bug_assert(left < right);
+
+    hull->left = left;
+    hull->right = GSL_MIN(
+        right + self->model.params.smc_k_coalescent.hull_offset, self->sequence_length);
+    hull->lineage = lineage;
+    hull->count = 0;
+    hull->insertion_order = UINT64_MAX;
+    tsk_bug_assert(lineage->prev == NULL);
+    lineage->hull = hull;
+
+out:
+    return hull;
+}
+
+static hullend_t *MSP_WARN_UNUSED
+msp_alloc_hullend(msp_t *self, double position, label_id_t label)
+{
+    hullend_t *hullend = NULL;
+
+    if (object_heap_empty(&self->hullend_heap[label])) {
```

Comment:
Same here, we're not exercising the expand logic here in the C tests (will probably happen automatically, though?)